### PR TITLE
[GSoC] Show Avg Ease in note mode

### PR DESCRIPTION
**Pull request by viciousAegis**
## Purpose / Description
Follow up to #12035. User can now switch to showing just one card per note, but, the columns still show data for the card alone. As mentioned in the Anki Docs [here](https://docs.ankiweb.net/browsing.html#columns), a few columns display slightly different data in notes mode. This PR tackles displaying avg ease.

## Approach
Create a function for a `Card` to return the average ease of its note, and if `inCardsMode` is not true, (i.e. browser is in notes mode), we return the avg ease to the column header.

## How Has This Been Tested?

Manually on mobile device, Unit test added as well

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by ShridharGoel — DISMISSED** [review]

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 1]
```diff
@@ -459,6 +459,27 @@ open class Card : Cloneable {
...
     } // In Anki Desktop, a card with oDue <> 0 && oDid == 0 is not marked as dynamic.
 
+    /**
+     * returns the average ease of the note of the card,
+     * or if all the cards in the note are new, returns -1
+     */
+    fun avgEaseOfNote(): Int {
+        var avgEase = 0
+        var oldCardCount = 0
+        val cards = this.note().cards().toMutableList()
```
[the remark this answers is not in the document]
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:469]: There is no such need, I see, will convert to just a list then 👍

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 2]
```diff
@@ -459,6 +459,27 @@ open class Card : Cloneable {
...
     } // In Anki Desktop, a card with oDue <> 0 && oDid == 0 is not marked as dynamic.
 
+    /**
+     * returns the average ease of the note of the card,
+     * or if all the cards in the note are new, returns -1
+     */
+    fun avgEaseOfNote(): Int {
+        var avgEase = 0
+        var oldCardCount = 0
+        val cards = this.note().cards().toList()
```
Comment (ShridharGoel) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:469]: Not even `toList()` was required, could have been left as `ArrayList`.

File: AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.kt [hunk 3]
```diff
@@ -337,4 +337,41 @@ class CardTest : RobolectricTest() {
...
+        }
+        // avg ease = (3000/10 + 1500/10 + 100/10 + 750/10) / 4 = [156.25] = 156
+        var avgEase = 156
+        assertEquals(avgEase, note.firstCard().avgEaseOfNote())
+
+        // test case: one card is new
+        i = 1
+        for (card in note.cards()) {
+            if (i == 3)
+                card.type = Consts.CARD_TYPE_NEW
```
Comment (ShridharGoel) [inline AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.kt:360]: According to Kotlin style guidelines, if you are not using brackets, then the statement should be on the same line as the condition.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 4]
```diff
@@ -459,6 +459,27 @@ open class Card : Cloneable {
...
+            oldCardCount++
+            avgEase += it.factor / 10
+        }
+
+        return if (oldCardCount == 0) {
+            -1
+        } else {
+            avgEase / oldCardCount
+        }
+    }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:481]: Design: This shouldn' be in `libAnki`, this should be in `com.ichi2.anki.servicelayer.CardService`

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 5]
```diff
@@ -459,6 +459,27 @@ open class Card : Cloneable {
...
+        var oldCardCount = 0
+        val cards = this.note().cards()
+
+        cards.filter { it.type != Consts.CARD_TYPE_NEW }.forEach {
+            oldCardCount++
+            avgEase += it.factor / 10
+        }
+
+        return if (oldCardCount == 0) {
+            -1
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:477]: Can we return `null` here instead?

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 6]
```diff
@@ -2623,6 +2627,14 @@ open class CardBrowser :
         @JvmField
         var sCardBrowserCard: Card? = null
 
+        /**
+         * Boolean that keeps track of whether the browser is working in
+         * Cards mode or Notes mode.
+         * True by default.
+         * */
+        @get:VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+        var inCardsMode: Boolean = true
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:2636]: Design: keep this non-global, it makes the app state much harder to reason about

File: AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.kt [hunk 7]
```diff
@@ -337,4 +337,40 @@ class CardTest : RobolectricTest() {
...
+        // factor for cards: 3000, 1500, 1000, 750
+        var i = 1
+        for (card in note.cards()) {
+            card.type = Consts.CARD_TYPE_REV
+            card.factor = 3000 / i++
+            card.flush()
+        }
+        // avg ease = (3000/10 + 1500/10 + 100/10 + 750/10) / 4 = [156.25] = 156
+        var avgEase = 156
+        assertEquals(avgEase, note.firstCard().avgEaseOfNote())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.kt:354]: Remove the variable and supply it directly in the assertion
```suggestion
        assertEquals(156, note.firstCard().avgEaseOfNote())
```

File: AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.kt [hunk 8]
```diff
@@ -337,4 +337,40 @@ class CardTest : RobolectricTest() {
...
         assertEquals("((filtered))", c.dueString)
     }
+
+    @Test
+    fun testAvgEaseOfNote() {
+        // basic case: no cards are new
+        val note = addNoteUsingModelName("Cloze", "{{c1::Hello}}{{c2::World}}{{c3::foo}}{{c4::bar}}", "extra")
+        // factor for cards: 3000, 1500, 1000, 750
+        var i = 1
+        for (card in note.cards()) {
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.kt:347]: you can use 

`for ((i, card) in note.cards().withIndex()) {` (i will be 0-based)

File: AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.kt [hunk 9]
```diff
@@ -337,4 +337,40 @@ class CardTest : RobolectricTest() {
...
+        var avgEase = 156
+        assertEquals(avgEase, note.firstCard().avgEaseOfNote())
+
+        // test case: one card is new
+        i = 1
+        for (card in note.cards()) {
+            if (i == 3) card.type = Consts.CARD_TYPE_NEW
+            card.flush()
+            i++
+        }
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/libanki/CardTest.kt:362]: Can this be changed to 

```kotlin
note.cards()[n].apply {
    type = Consts.CARD_TYPE_NEW
    flush()
}
```

**Review by david-allison — CHANGES_REQUESTED** [review]
I like this PR and agree with the outcome, I do have some minor nitpicks on code quality, which are 'implementer's choice', but shouldn't be too time-consuming. 

My main request is to keep the `inCardsMode` non-global (even if this requires passing in a lambda to `CardCache`). We can talk design if necessary.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 10]
```diff
@@ -459,6 +459,27 @@ open class Card : Cloneable {
...
+        } else {
+            avgEase / oldCardCount
+        }
+    }
+
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:482]: As mentioned in Discord, I don't like code rewrites but I feel this shows a more readable and more succinct method of doing this

```kotlin
        val nonNewCards = note().cards().filter { it.type != Consts.CARD_TYPE_NEW }
        if (nonNewCards.isEmpty()) {
            return null
        }
        return nonNewCards.map { it.factor / 10 }.average().toInt()
```

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:482]: I'm sorry but I'm not fine with this for multiple reasons.

If we decide to call `note()` then i believe we should put the core of the method in `Notes.kt` and just call it. `fun avgEaseOfNote() = note().avgEase()`.

Also, I believe `/10` should be outside of the map. Not because it will ensure we divide a single time. But mostly because we are losing the last digit of precision each time. Instead, we can average first with best data we have and then remove the last digit.
Except if you followed upstream code. In this case, I'd really want you to document, probably in code, at least in commit message, the name of the upstream function you are porting, so that I can go compare myself.

Implementor choice:
I prefer 
```Kotlin
return 
    if (nonNewCards.isEmpty()) {
        null
    } else {
      nonNewCards.map { it.factor / 10 }.average().toInt()
  }
```
I'd even consider calling `average` directly and check whether it returns Double.NaN, but it does not seems to add any readability.

Finally, and most important, this makes a query for the note and one for each card. This means that this computation will be quite slow, especially once you put it in a list of cards. 
I believe in this case, a direct sqlite query would be justified. Good news is that `AVG` returns null on an empty set.

Admittedly, the real ideal would be to be able to get all required informations simultaneously with the findNote queries, but that would require so much extra work (and probably must be done through Rust) that I clearly I won't block on this, and won't require this in the foreseeable future

Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:482]: >Also, I believe /10 should be outside of the map. Not because it will ensure we divide a single time. But mostly because we are losing the last digit of precision each time. Instead, we can average first with best data we have and then remove the last digit.

I just followed the implementation of ease that is given here. Although I can't find the upstream equivalent, on manually checking the avg ease on desktop and on ankidroid, they were exactly the same. I'll check again with `/ 10` outside the map, and if the results are the same, I'll apply that

https://github.com/ankidroid/Anki-Android/blob/c744c99c41f02f7f6b476c1d4cacf58ec6f86eff/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt#L2322-L2328

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt (hunk 4 above)
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:481]: Putting it in `Note.kt` as suggested by Arthur, is that fine?

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:481]: `NoteService.kt`. It's not part of `libAnki`

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 6 above)
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:2636]: What would I have to do to pass this value to `Cache` then?

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt [hunk 11]
```diff
@@ -459,6 +460,8 @@ open class Card : Cloneable {
         return LanguageUtil.getShortDateFormatFromS(date)
     } // In Anki Desktop, a card with oDue <> 0 && oDid == 0 is not marked as dynamic.
 
+    fun avgEaseOfNote() = avgEase(this.note())
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt:463]: Implementor choice:
`avgEase(note())`

**Review by Arthur-Milchior — APPROVED** [review]

File: AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt (hunk 10 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Card.kt]: This explain how we got the value for a single card. My question was really about how upstream compute the average.
But I understand that it may be a hard question since it requires to navigate the upstream code, potentially in rust

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.kt [hunk 12]
```diff
@@ -198,6 +199,20 @@ object NoteService {
...
+     * returns the average ease of all the non-new cards in the note,
+     * or if all the cards in the note are new, returns null
+     */
+    fun avgEase(note: Note): Int? {
+        val nonNewCards = note.cards().filter { it.type != Consts.CARD_TYPE_NEW }
+
+        return if (nonNewCards.isEmpty()) {
+            null
+        } else {
+            (nonNewCards.map { it.factor }.average() / 10).toInt()
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.kt:212]: I’d still love to get at least a TODO asking to do it in sql directly if you don’t feel doing it here

Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.kt:212]: Got it 👍

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt [hunk 13]
```diff
@@ -500,7 +500,7 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
                 Timber.d("The search found %d notes", searchResult_.size)
                 var position = 0
                 for (nid in searchResult_) {
-                    val card = CardCache(Note(col, nid).firstCard().id, col, position++)
+                    val card = CardCache(Note(col, nid).firstCard().id, col, position++, inCardMode = { false })
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:503]: @Arthur-Milchior I believe I've overcomplicated this and we can go with a `Boolean`.

I presume the cache is cleared and a new search is performed when the note -> card switch is made?

Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:503]: Yes, a new search is performed every time the switch is made

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:503]: The question is probably for @viciousAegis not for me. Vicious Aegis knows the code far better than me.
I agree that I don’t understand the need for the function here, so I’m totally fine with a simple Boolean.

I don’t understand how YOU had overcomplicated things, but I guess I miss a suggestion you had made about it

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.kt [hunk 14]
```diff
@@ -198,6 +199,21 @@ object NoteService {
...
+     * returns the average ease of all the non-new cards in the note,
+     * or if all the cards in the note are new, returns null
+     */
+    fun avgEase(note: Note): Int? {
+        val nonNewCards = note.cards().filter { it.type != Consts.CARD_TYPE_NEW }
+
+        return nonNewCards.average { it.factor }?.let { it / 10 }?.toInt()
+    }
+
+    fun <T> Iterable<T>.average(f: (T) -> Int): Double? {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.kt:213]: This is not related to `note` at all. So if we decide to use it, I believe it should be in some "util" class somewhere.

**Review by Arthur-Milchior — APPROVED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.kt (hunk 14 above)
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.kt:213]: is `CollectionsUtil` a good place for this?

**Review by david-allison — APPROVED** [review]
Beauty!
